### ISSUE TITLE
Exclusion of kube-system in EndpointsWatcher leads to broken DNS resolution

### ISSUE THREAD

**Issue by JacobHenner** [linkerd/linkerd2#8721]
### What is the issue?

On several occasions and in several clusters, I've encountered an issue where DNS resolution begins to fail for linkerd-enabled pods, and does not recover until the cluster's linkerd-destination pods are replaced with new destination pods (e.g. `kubectl rollout restart -n linkerd deployment/linkerd-destination`). The issue is limited to DNS over TCP, which is used frequently in my environment.

An investigation revealed that these DNS resolution failures occurred because the queries were being sent (by linkerd-proxy) to endpoints that were no longer available, instead of to live endpoints. `linkerd diagnostics endpoints kube-dns.kube-system.svc.cluster.local:53` would also display endpoints that were no longer available instead of the live endpoints. `kubectl get endpoints` and `kubectl get endpointslices` produced the expected results, with up-to-date addresses.

A look through the codebase suggests this occurs because the watchers in EndpointsWatcher explicitly ignore events in kube-system. Changes in endpoint IP addresses are not detected, leading linkerd-destination to serve stale endpoint data if/when endpoints change (e.g. a CoreDNS pod is replaced by another).

### How can it be reproduced?

* Create a standard Kubernetes cluster
* Install linkerd
* Ensure there is continuous TCP DNS request load in the cluster (from linkerd-enabled pods) to ensure linkerd-destination does not drop the portPublisher that is in use.
* Confirm that the kube-dns endpoints (or endpointslices) are up-to-date: `kubectl get -n kube-system endpoints kube-dns`, and compare to the output of `linkerd diagnostics endpoints kube-dns.kube-system.svc.cluster.local:53`. Note that they correspond properly.
* Perform a rolling restart of CoreDNS: `kubectl rollout restart -n kube-system deployment/coredns`
* Re-examine the endpoints reported by Kubernetes, as well as the endpoints reported by linkerd. Note that they have diverged, and that linkerd is still reporting the previously-reported entries instead of the newly-created pod endpoints.
* Confirm DNS resolution is broken by attempting a TCP DNS query in a linkerd-enabled pod.

### Logs, error output, etc

No specific logs or error output illustrate the issue. However, debug logs of the linkerd-destination pods demonstrate that kube-system endpoints/services are not updated by watcher, as opposed to endpoints in other namespaces that have corresponding log lines for updates.

### output of `linkerd check -o short`

```text
Linkerd core checks
===================

linkerd-existence
-----------------
‼ cluster networks can be verified
    the following nodes do not expose a podCIDR:
        (redacted list of nodes)
    see https://linkerd.io/2/checks/#l5d-cluster-networks-verified for hints

linkerd-version
---------------
‼ cli is up-to-date
    is running version 22.5.1 but the latest edge version is 22.6.2
    see https://linkerd.io/2/checks/#l5d-version-cli for hints

control-plane-version
---------------------
‼ control plane is up-to-date
    is running version 22.5.1 but the latest edge version is 22.6.2
    see https://linkerd.io/2/checks/#l5d-version-control for hints

linkerd-control-plane-proxy
---------------------------
‼ control plane proxies are up-to-date
    some proxies are not running the current version:
        * linkerd-destination-5678786dbc-55bqd (edge-22.5.1)
        * linkerd-destination-5678786dbc-f66t8 (edge-22.5.1)
        * linkerd-destination-5678786dbc-k5rsr (edge-22.5.1)
        * linkerd-identity-749478f779-4trf4 (edge-22.5.1)
        * linkerd-identity-749478f779-5vfzl (edge-22.5.1)
        * linkerd-identity-749478f779-tmtvf (edge-22.5.1)
        * linkerd-proxy-injector-7f9cc5d964-9hj95 (edge-22.5.1)
        * linkerd-proxy-injector-7f9cc5d964-nfgc5 (edge-22.5.1)
        * linkerd-proxy-injector-7f9cc5d964-pf95d (edge-22.5.1)
    see https://linkerd.io/2/checks/#l5d-cp-proxy-version for hints

Status check results are √
```

### Environment

* Kubernetes version: 1.21
* Cluster environment: EKS
* Host OS: Amazon Linux
* Linkerd version: edge-22.5.1

### Possible solution

Remove watcher exclusions for kube-system updates. This [has been proposed before](https://github.com/linkerd/linkerd2/pull/4133), but was dropped due to inactivity. I've opened #8720 for this proposed change.

There was [a concern](https://github.com/linkerd/linkerd2/pull/4133#issuecomment-594983588) that frequent updates of kube-system Endpoints [used for leader elections](https://github.com/kubernetes/kubernetes/issues/86286) would need to be special-cased, and it was suggested that this was why kube-system had been excluded originally. However, these leader elections [use Leases instead of Endpoints](https://github.com/kubernetes/kubernetes/pull/94603) by default since Kubernetes 1.20, obviating this concern. When #4133 was considered, Endpoints were still in use as the default leader election mechanism.

### Additional context

Upon initial review, one might wonder "if this issue affects DNS resolution, why haven't more users complained?". 

For this issue to occur, all of the following conditions must be true:

* DNS is using TCP (as opposed to UDP, which is more typical).
* The endpoints for the DNS service have changed since linkerd-destination last created its internal representation of the service's endpoints.
* The listener count for the portPublisher must not drop to 0 (since this would cause `unsubscribe()` to drop the outdated internal representation, leading to the generation of a current representation on the next request for the service).

One might also wonder why this issue isn't as likely to be observed for other services in kube-system:

* Generally, there are few linkerd-enabled consumers of kube-system services. DNS is an obvious exception.
* Since DNS is queried frequently, it's less likely the portPublisher listener count drops to 0. Other services in kube-system are more likely to see their listener count drop to 0, which would effectively result in their endpoints being updated on the next request.

### Would you like to work on fixing this bug?

yes

**Comment by JacobHenner** [comment]
* #4133 was proposed in response to https://github.com/linkerd/linkerd2/issues/4055. #4055 described difficulties with linkerd-enabled services in kube-system. The issue described here is not about linkerd-enabled services in kube-system, but rather connections to non-meshed services in kube-system _from_ linkerd-enabled pods in other namespaces.
* When the kube-system exclusion was considered back in 2020, one contributor believed [it would be safe to remove](https://github.com/linkerd/linkerd2/issues/4055#issuecomment-591001319).

**Comment by JacobHenner** [comment]
See https://github.com/linkerd/linkerd2/pull/8720 for proposed solution.

**Comment by olix0r** [comment]
@JacobHenner Thanks for the detailed explanation and PR! At a first glance, this all makes sense to me. We'll take a look at this as a team and let you know if we have any followup questions.

**Comment by kleimkuhler** [comment]
This was closed by #8720
